storage_service/locations/forms.py

```diff
@@ -392,10 +394,120 @@ def __init__(self, *args, **kwargs):
         self.fields["status_reason"].required = True
 
 
+class HeaderWidget(forms.MultiWidget):
+    """Key/value text inputs widget for Callback header."""
+
+    def __init__(self, *args, **kwargs):
+        widgets = (forms.TextInput(), forms.TextInput())
+        super(HeaderWidget, self).__init__(widgets, *args, **kwargs)
+
+    def decompress(self, value):
+        return value or [None, None]
+
+
+class HeaderField(forms.fields.MultiValueField):
+    """Key/value form field for Callback header."""
+
+    widget = HeaderWidget
+
+    def __init__(self, *args, **kwargs):
+        fields = (forms.fields.CharField(), forms.fields.CharField())
+        super(HeaderField, self).__init__(fields, *args, **kwargs)
+
+    def compress(self, data_list):
+        return data_list or [None, None]
+
+
 class CallbackForm(forms.ModelForm):
     class Meta:
         model = models.Callback
-        fields = ("uri", "event", "method", "expected_status")
+        fields = ("event", "uri", "method", "body", "expected_status", "enabled")
+
+    def __init__(self, *args, **kwargs):
+        """Extends model form initialization.
+
+        Adds extra fields for Callback headers and reorders form fields.
+        """
+        super(CallbackForm, self).__init__(*args, **kwargs)
+        # Save field order to include the headers after the method field
+        field_order = ["event", "uri", "method"]
+        # Add fields for existing headers
+        counter = 0
+        if self.instance:
+            for header in self.instance.get_headers().items():
```

Comment:
It's been a while since I've used one, but [formsets](https://docs.djangoproject.com/en/1.8/topics/forms/formsets/) exist for this use case — although it may not be that easy because headers isn't a model. You should be able to use a separate HeaderForm and formset for that, though.

Formsets require some JS handling (not really more than you already have in place though), it looks like it hasn't been updated in a while but https://github.com/elo80ka/django-dynamic-formset may make that easier.

Reply:
I looked at formsets and examples using django-dynamic-fomset but I thought it would be easier to treat the header with a dynamic multi-input. In the end I wasn't so sure about that, but it was kind of late to go back :(

Reply:
Ok, sounds good.